src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +381,181 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
                 segments.Add(writer.ToSourceText());
             }
         }
+
+        /// <summary>
+        /// Delegates to SourceTexts within the CompositeText to determine line information.
+        /// </summary>
+        private sealed class CompositeTextLineInfo : TextLineCollection
+        {
+            private readonly CompositeText _compositeText;
+
+            /// <summary>
+            /// The starting line number for the correspondingly indexed SourceTexts in _compositeText.Segments.
+            /// Multiple consecutive entries could indicate the same line number if the corresponding
+            /// segments don't contain newline characters. 
+            /// </summary>
+            /// <remarks>
+            /// This will be of the same length as _compositeText.Segments
+            /// </remarks>
+            private readonly ImmutableArray<int> _segmentLineNumbers;
+
+            // The total number of lines in our _compositeText
+            private readonly int _lineCount;
+
+            public CompositeTextLineInfo(CompositeText compositeText)
+            {
+                var segmentLineNumbers = ArrayBuilder<int>.GetInstance();
+                var accumulatedLineCount = 0;
+
+                Debug.Assert(compositeText.Segments.Length > 0);
```

Comment:
consider making a contracct call.  i like contracts for O(1) checks, and Debug asserts for more expenisve ones.

Reply:
I've converted some of the Debug.Asserts to throw, but others I've kept as is as it feels like the exception that would be thrown is just as good as the one that I would add.